File: tests/fixtures/csharp_rules/Shop.cs
// Fixture exercising every v1 architectural-rule tag (C#).
using System.Collections.Generic;
using CodeConstraints.Rules;

namespace Shop
{
    [Layer("domain")]
    [Sealed]
    public class Repository
    {
        public object Find(int key) => null;
    }

    public class SpecialRepository : Repository { }  // sealed VIOLATION

    [Immutable]
    public class Money
    {
        private int amount;

        public Money(int amount) { this.amount = amount; }

        public int Add(int other) => this.amount + other;  // clean

        public void Reset() { this.amount = 0; }  // immutable VIOLATION
    }

    public class OrderService
    {
        [NoInstantiation(Allow = new[] { "List" })]
        public List<int> Collect()
        {
            var acc = new List<int>();  // allowed by Allow
            return acc;
        }

        [NoInstantiation]
        public int Total(List<int> items) => 0;  // clean
    }

    [Factory(Creates = new[] { "Repository" })]
    public class RepositoryFactory
    {
        public Repository Create() => new Repository();  // allowed: designated factory
    }

    public class Sneaky
    {
        public Repository Make() => new Repository();  // factory VIOLATION
    }
}

File: tests/fixtures/python_rules/shop.py
"""Fixture exercising every v1 architectural-rule tag (Python)."""

from cdec_rules import factory, immutable, layer, no_instantiation, sealed


@layer("domain")
@sealed
class Repository:
    """A sealed domain class."""

    def find(self, key: int):
        return None


class SpecialRepository(Repository):  # sealed VIOLATION: subclasses a @sealed class
    pass


@immutable
class Money:
    def __init__(self, amount: int):
        self.amount = amount

    def add(self, other: int) -> int:
        return self.amount + other  # clean: reads only

    def reset(self) -> None:
        self.amount = 0  # immutable VIOLATION: reassigns a field outside __init__


class OrderService:
    @no_instantiation
    def total(self, items: list) -> int:
        return sum(items)  # clean: no construction

    @no_instantiation(allow=["list"])
    def collect(self):
        acc = list()  # allowed by `allow`
        return acc


@factory(creates=["Repository"])
class RepositoryFactory:
    def create(self) -> Repository:
        return Repository()  # allowed: this is the designated factory


class Sneaky:
    def make(self):
        return Repository()  # factory VIOLATION: constructs Repository outside the factory
